modifying list_scan to accept multiple motors

awalter-bnl opened the pull request:
modifying list_scan to accept multiple motor/list pairs

## Description
Current API:
```
list_scan(dets,` motor, [...])  # supports one motor only
scan(dets, motor, start, stop, motor2, start2, stop2, num_steps)  # "inner product" over motors
```
Proposed API:
```
list_scan(dets, motor, [...], motor2, [...])  # If list lengths are unequal, blow up.
list_grid_scan(dets, motor, [...], motor2, [...], motor3, [...], snaking=True|False|tuple)
```
## Motivation and Context
fixes #1144 

## How Has This Been Tested?
has not yet been tested

danielballan commented on bluesky/plans.py:
```diff
@@ -90,36 +96,67 @@ def list_scan(detectors, motor, steps, *, per_step=None, md=None):
+    motors = []
+    pos_lists = []
+    for motor, pos_list in partition(2,args):
+        lengths[motor.name] = len(pos_list)
```
```suggestion
        pos_list = list(pos_list)  # Ensure list (accepts any finite iterable).
        lengths[motor.name] = len(pos_list)
```

danielballan commented on bluesky/plans.py:
```diff
@@ -90,36 +96,67 @@ def list_scan(detectors, motor, steps, *, per_step=None, md=None):
+        lengths[motor.name] = len(pos_list)
+        motors.append(motor)
+        pos_lists.append(pos_list)
+    lengths = {motor.name: len(pos_list) for motor, pos_list
```
I guess this statement is redundant to the loop above.

awalter-bnl replied:
yes, thanks for highlighting. I have removed it

danielballan commented on bluesky/plans.py:
```diff
@@ -90,36 +96,67 @@ def list_scan(detectors, motor, steps, *, per_step=None, md=None):
+
+    if not length_check:
+        raise ValueError("The lengths of all lists in *args must be the same"
+                         f"However the lengths in args are : {lengths}")
```
A+ error message!

danielballan commented on bluesky/plans.py:
```diff
@@ -90,36 +96,67 @@ def list_scan(detectors, motor, steps, *, per_step=None, md=None):
+                      for motor, pos in zip(motors, pos_lists)})
 
-    return (yield from inner_list_scan())
+    return (yield from scan_nd(detectors, cycler, per_step=per_step, md=_md))
```
This should be a `cycler` object, not a list of `cycler` objects. I believe that above , where you define the cycler, you want something like:

```python
cycler = sum(cycler(motor, pos_list) for motor, pos_list in partition(2,args))
```

awalter-bnl replied:
thanks I have made this change

danielballan commented on bluesky/plans.py:
```diff
@@ -90,36 +96,67 @@ def list_scan(detectors, motor, steps, *, per_step=None, md=None):
+    if not length_check:
+        raise ValueError("The lengths of all lists in *args must be the same"
+                         f"However the lengths in args are : {lengths}")
+
```
At this point, where we know we have one consistent length, it might be worth defining a variable `length = len(post_lists([0])` to improve readability, since you use this expression several places below.

awalter-bnl replied:
done, good call

awalter-bnl commented:
I have added `list_grid_scan`, `rel_list_scan` and `rel_list_grid_scan` to the PR.

I have also modified them to use `bluesky.plan_patterns` to mirror the approach used on `scan` and `grid_scan`.

Finally I have got the existing tests passing locally. Feel free to leave comments, I will work on adding tests of the new functionality.

danielballan commented on bluesky/plans.py:
```diff
@@ -148,14 +198,147 @@ def rel_list_scan(detectors, motor, steps, *, per_step=None, md=None):
     return (yield from inner_relative_list_scan())
 
 
+def list_grid_scan(detectors, *args, per_step=None, md=None):
```
The `snake_axes` parameter is missing.

awalter-bnl replied:
fixed

danielballan commented on bluesky/plans.py:
```diff
@@ -148,14 +198,147 @@ def rel_list_scan(detectors, motor, steps, *, per_step=None, md=None):
+        can be any 'settable' object (motor, temp controller, etc.).
+
+    snake_axes
+        which axes should be snaked, either None (do not snake any axes),
```
We settled on `False` for this, though if the user passes in `None` it will have the same result as `False`. See [comment](https://github.com/bluesky/bluesky/issues/1189#issuecomment-492872715) and discussion below. (The default of `False`  well suggests to the rushed reader that `True` might be a valid option.)

awalter-bnl replied:
fixed

danielballan commented on bluesky/plans.py:
```diff
@@ -148,14 +198,147 @@ def rel_list_scan(detectors, motor, steps, *, per_step=None, md=None):
+        are a list of positions, all lists must have the same length. Motors
+        can be any 'settable' object (motor, temp controller, etc.).
+
+    snake_axes
```
```suggestion
    snake_axes : boolean or Iterable, optional
```

awalter-bnl replied:
fixed

awalter-bnl commented:
thanks for the quick review turnaround @danielballan I have made the suggested changes

danielballan commented on bluesky/plans.py:
```diff
@@ -334,8 +334,8 @@ def rel_list_grid_scan(detectors, *args, per_step=None, md=None):
     def inner_relative_list_grid_scan():
-        return (yield from list_grid_scan(detectors, *args, per_step=per_step,
-                                          md=_md))
+        return (yield from list_grid_scan(detectors, *args, snake_axes,
```
Any parameters after `*args` must be passed by keyword (as you'll discovered when you write a test for this).

awalter-bnl replied:
ok thanks, this will be fixed in the next commit !-)

awalter-bnl commented:
added tests of `list_scan` and `rel_list_scan` test of `list_grid_scan` and `rel_list_grid_scan` to follow

awalter-bnl commented:
All coding and tests are complete, @stuartcampbell is there a process or set of guidelines for moving these to 'ready to review' status or is it when the author thinks it is ready?

danielballan commented on bluesky/plan_patterns.py:
```diff
@@ -269,6 +269,90 @@ def spiral_fermat(x_motor, y_motor, x_start, y_start, x_range, y_range, dr,
+
+    Parameters
+    ----------
+    args : list of {Positioner, position_list}
```
`of {Positioner, position_list}` gives me the impression of a nested structure. Would it be better to just say `list`? I'm sure mileage will very here, and I could be in the minority.

awalter-bnl replied:
fixed

danielballan commented on bluesky/plan_patterns.py:
```diff
@@ -269,6 +269,90 @@ def spiral_fermat(x_motor, y_motor, x_start, y_start, x_range, y_range, dr,
+                snaking.append(False)
+        else:
+            raise ValueError('The snake_axes arg to ``outer_list_pattern`` '
+                             'must be either None (do not snake any axes), '
```
```suggestion
                             'must be either False (do not snake any axes), '
```

awalter-bnl replied:
fixed

danielballan commented on bluesky/plan_patterns.py:
```diff
@@ -269,6 +269,90 @@ def spiral_fermat(x_motor, y_motor, x_start, y_start, x_range, y_range, dr,
+        print (motor)
+        if not snake_axes:
+            snaking.append(False)
+        elif snake_axes is True:
```
```suggestion
        elif snake_axes:
```

awalter-bnl replied:
So in this case it is necessary to explicitly state `is True`, otherwise a non empty `list` would pass this condition and we need those to be handled by the next `elif` block.

Does this make sense to you @danielballan ?

danielballan replied:
Oh, I didn't think of that. In that case can we check for the `Iterable` first, and then accept anything that is "truth-y" but not `Iterable` as equivalent to `True`?

awalter-bnl replied:
done

danielballan commented on bluesky/plan_patterns.py:
```diff
@@ -269,6 +269,90 @@ def spiral_fermat(x_motor, y_motor, x_start, y_start, x_range, y_range, dr,
+                snaking.append(False)
+            else:
+                snaking.append(True)
+        elif type(snake_axes) is list:
```
Please use `isinstance(snake_axes, collections.abc.Iterable)` to accept _any_ iterable, including tuples. dict keys, types we haven't heard of, etc.

awalter-bnl replied:
fixed

danielballan commented on bluesky/plan_patterns.py:
```diff
@@ -269,6 +269,90 @@ def spiral_fermat(x_motor, y_motor, x_start, y_start, x_range, y_range, dr,
+    snaking = []
+    cyclers = []
+    for motor, pos_list in partition(2, args):
+        print (motor)
```
Stray print

awalter-bnl replied:
removed

danielballan commented on bluesky/plans.py:
```diff
@@ -90,46 +96,90 @@ def list_scan(detectors, motor, steps, *, per_step=None, md=None):
     :func:`bluesky.plans.rel_list_scan`
     """
+    if len(args) % 2 != 0:
+            raise ValueError("The list of arguments must contain a list of"
```
```suggestion
            raise ValueError("The list of arguments must contain a list of "
```

awalter-bnl replied:
fixed

awalter-bnl commented:
Thanks again to @danielballan for the rapid fire reviews, there is one outstanding comment above to which I have responded all others have been resolved as per the suggestions.

awalter-bnl commented:
@danielballan I have resolved all your suggestions, it is ready for another look.

awalter-bnl commented:
I have not yet released it as “ready for review” as I am waiting on feedback from @stuartcampbell regarding the appropriate process to do that, but yes I am happy to have other look at it.

awalter-bnl commented:
I have moved this from 'draft' to 'ready to review'. @danielballan it appears you are still requesting changes, I think based on your last comment that I have resolved all of the outstanding issues can I ask you to check?

danielballan commented:
Power-cycled to trigger CI

danielballan commented:
This still looks overall good. I removed my "Approval" until the errors are resolved, but I think as soon as this passes it should be fine after one more read.

awalter-bnl commented:
finally got this to pass and build docs. it is ready to go

mrakitin commented on bluesky/plan_patterns.py:
```diff
@@ -269,6 +270,89 @@ def spiral_fermat(x_motor, y_motor, x_start, y_start, x_range, y_range, dr,
+        patterned like (``motor1, position_list1,``
+                        ``...,``
+                        ``motorN, position_listN``)
+        Motors can be any 'setable' object (motor, temp controller, etc.)
```
```suggestion
        Motors can be any 'settable' object (motor, temp controller, etc.)
```

mrakitin commented on bluesky/plan_patterns.py:
```diff
@@ -269,6 +270,89 @@ def spiral_fermat(x_motor, y_motor, x_start, y_start, x_range, y_range, dr,
+                        ``...,``
+                        ``motorN, position_listN``)
+        Motors can be any 'setable' object (motor, temp controller, etc.)
+        ``position_list``'s are a list of postiions, all lists must have the
```
I think it will be more precise:
```suggestion
        ``position_list``'s are lists of positions, all lists must have the
```

mrakitin commented on bluesky/plan_patterns.py:
```diff
@@ -269,6 +270,89 @@ def spiral_fermat(x_motor, y_motor, x_start, y_start, x_range, y_range, dr,
+    cyc : cycler
+    '''
+    if len(args) % 2 != 0:
+        raise ValueError("wrong number of positional arguments")
```
```suggestion
        raise ValueError("Wrong number of positional arguments for 'inner_list_product'")
```

mrakitin commented on bluesky/plan_patterns.py:
```diff
@@ -269,6 +270,89 @@ def spiral_fermat(x_motor, y_motor, x_start, y_start, x_range, y_range, dr,
+                        ``motorN, position_listN``)
+
+        The first motor is the "slowest", the outer loop. ``position_list``'s
+        are a list of positions, all lists must have the same length.
```
```suggestion
        are lists of positions, all lists must have the same length.
```

mrakitin commented on bluesky/plan_patterns.py:
```diff
@@ -269,6 +270,89 @@ def spiral_fermat(x_motor, y_motor, x_start, y_start, x_range, y_range, dr,
+        are a list of positions, all lists must have the same length.
+.
+    snake_axes
+        which axes should be snaked, either None (do not snake any axes),
```
```suggestion
        which axes should be snaked, either ``None`` (do not snake any axes),
```

mrakitin commented on bluesky/plan_patterns.py:
```diff
@@ -269,6 +270,89 @@ def spiral_fermat(x_motor, y_motor, x_start, y_start, x_range, y_range, dr,
+.
+    snake_axes
+        which axes should be snaked, either None (do not snake any axes),
+        True (snake all axes) or a list of axes to snake. 'snaking an axis'
```
```suggestion
        ``True`` (snake all axes) or a list of axes to snake. 'snaking an axis'
```

mrakitin commented on bluesky/plan_patterns.py:
```diff
@@ -269,6 +270,89 @@ def spiral_fermat(x_motor, y_motor, x_start, y_start, x_range, y_range, dr,
+            else:
+                snaking.append(True)
+        else:
+            raise ValueError('The snake_axes arg to ``outer_list_pattern`` '
```
Should it be `outer_list_product`?
```suggestion
            raise ValueError('The snake_axes arg to ``outer_list_product`` '
```

mrakitin commented on bluesky/plans.py:
```diff
@@ -90,46 +99,94 @@ def list_scan(detectors, motor, steps, *, per_step=None, md=None):
+    motors = []
+    pos_lists = []
+    length = None
+    for motor, pos_list in partition(2,args):
```
```suggestion
    for motor, pos_list in partition(2, args):
```

mrakitin commented on bluesky/plans.py:
```diff
@@ -90,46 +99,94 @@ def list_scan(detectors, motor, steps, *, per_step=None, md=None):
+                       list(lengths.values()))
+
+    if not length_check:
+        raise ValueError("The lengths of all lists in *args must be the same"
```
```suggestion
        raise ValueError("The lengths of all lists in *args must be the same. "
```

mrakitin commented on bluesky/plans.py:
```diff
@@ -148,14 +205,150 @@ def rel_list_scan(detectors, motor, steps, *, per_step=None, md=None):
-        return (yield from list_scan(detectors, motor, steps,
-                                     per_step=per_step, md=_md))
+        return (yield from list_scan(detectors, *args, per_step=per_step,
+                                     md=_md))
```
`md` can be on the same line as well.

mrakitin commented on bluesky/plans.py:
```diff
@@ -148,14 +205,150 @@ def rel_list_scan(detectors, motor, steps, *, per_step=None, md=None):
+                        ``motorN, position_listN``)
+
+        The first motor is the "slowest", the outer loop. ``position_list``'s
+        are a list of positions, all lists must have the same length. Motors
```
```suggestion
        are lists of positions, all lists must have the same length. Motors
```

mrakitin commented on bluesky/plans.py:
```diff
@@ -148,14 +205,150 @@ def rel_list_scan(detectors, motor, steps, *, per_step=None, md=None):
+        can be any 'settable' object (motor, temp controller, etc.).
+
+    snake_axes : boolean or iterable, optional
+        which axes should be snaked, either False (do not snake any axes),
```
```suggestion
        which axes should be snaked, either ``False`` (do not snake any axes),
```

mrakitin commented on bluesky/plans.py:
```diff
@@ -148,14 +205,150 @@ def rel_list_scan(detectors, motor, steps, *, per_step=None, md=None):
+
+    snake_axes : boolean or iterable, optional
+        which axes should be snaked, either False (do not snake any axes),
+        True (snake all axes) or a list of axes to snake. 'snaking an axis'
```
```suggestion
        ``True`` (snake all axes) or a list of axes to snake. 'snaking an axis'
```

mrakitin commented on bluesky/plans.py:
```diff
@@ -148,14 +205,150 @@ def rel_list_scan(detectors, motor, steps, *, per_step=None, md=None):
+
+    per_step : callable, optional
+        hook for customizing action of inner loop (messages per step)
+        See docstring of bluesky.plan_stubs.one_nd_step (the default) for
```
I guess it can be something like
```
:func:`bluesky.plan_stubs.one_nd_step`
```

mrakitin commented on bluesky/plans.py:
```diff
@@ -148,14 +205,150 @@ def rel_list_scan(detectors, motor, steps, *, per_step=None, md=None):
+           'plan_name': 'list_grid_scan',
+           'plan_pattern': 'outer_list_product',
+           'plan_pattern_args': dict(args=md_args).update(
+                {'snake_axes':snake_axes}),
```
```suggestion
                {'snake_axes': snake_axes}),
```

mrakitin commented on bluesky/plans.py:
```diff
@@ -148,14 +205,150 @@ def rel_list_scan(detectors, motor, steps, *, per_step=None, md=None):
+def rel_list_grid_scan(detectors, *args, snake_axes=False, per_step=None,
+                       md=None):
+    """
+    Scan over a mesh; each motor is on an independent trajectory.Each point is
```
```suggestion
    Scan over a mesh; each motor is on an independent trajectory. Each point is
```

mrakitin commented on bluesky/plans.py:
```diff
@@ -148,14 +205,150 @@ def rel_list_scan(detectors, motor, steps, *, per_step=None, md=None):
+        can be any 'settable' object (motor, temp controller, etc.).
+
+    snake_axes : boolean or Iterable, optional
+        which axes should be snaked, either False (do not snake any axes),
```
```suggestion
        which axes should be snaked, either ``False`` (do not snake any axes),
```

mrakitin commented on bluesky/plans.py:
```diff
@@ -148,14 +205,150 @@ def rel_list_scan(detectors, motor, steps, *, per_step=None, md=None):
+
+    snake_axes : boolean or Iterable, optional
+        which axes should be snaked, either False (do not snake any axes),
+        True (snake all axes) or a list of axes to snake. 'snaking an axis'
```
```suggestion
        ``True`` (snake all axes) or a list of axes to snake. 'snaking an axis'
```

mrakitin commented on bluesky/plans.py:
```diff
@@ -148,14 +205,150 @@ def rel_list_scan(detectors, motor, steps, *, per_step=None, md=None):
+        simple left-to-right trajectory.
+
+    per_step : callable, optional
+        hook for customizing action of inner loop (messages per step)
```
```suggestion
        hook for customizing action of inner loop (messages per step).
```

mrakitin commented on bluesky/plans.py:
```diff
@@ -148,14 +205,150 @@ def rel_list_scan(detectors, motor, steps, *, per_step=None, md=None):
+
+    per_step : callable, optional
+        hook for customizing action of inner loop (messages per step)
+        See docstring of bluesky.plan_stubs.one_nd_step (the default) for
```
```suggestion
        See docstring of :func:`bluesky.plan_stubs.one_nd_step` (the default) for
```

mrakitin commented on bluesky/plans.py:
```diff
@@ -148,14 +205,150 @@ def rel_list_scan(detectors, motor, steps, *, per_step=None, md=None):
+    :func:`bluesky.plans.rel_list_scan`
+     See Also
+    --------
+    :func:`bluesky.plans.list_scan`
```
These blocks are repeating.

awalter-bnl replied:
fixed

mrakitin commented on bluesky/plans.py:
```diff
@@ -148,14 +205,150 @@ def rel_list_scan(detectors, motor, steps, *, per_step=None, md=None):
+
+    motors= [motor for motor, pos_list in partition(2, args)]
+
+
```
Should be 1 line break within a function.

mrakitin commented on bluesky/tests/test_scans.py:
```diff
@@ -166,13 +166,43 @@ def test_ascan(RE, hw):
+         'motor1': 8, 'motor1_setpoint': 8}]
+
+    scan = bp.list_scan([hw.det], hw.motor1, [6, 7, 8],
+                        hw.motor2, [18, 28, 38])
```
Why not to have numpy-arrays as inputs for this case as in the tests below?

awalter-bnl replied:
good catch, I test with lists in another test this was suppose to test with numpy arrays.

mrakitin commented on doc/source/plans.rst:
```diff
@@ -219,6 +221,11 @@ incorporating these trajectories, use our general N-dimensional scan plan,
    grid_scan
    rel_grid_scan
    scan_nd
+   list_scan
```
Should the `list_scan` plan be removed from the previous block?

awalter-bnl replied:
I think they belong in both, as they now scan over 1 _or_ more dimensions so are both 1 dimensional scans _and_ multi-dimensional scans.

mrakitin replied:
OK, sounds good.

awalter-bnl commented:
@mrakitin thanks for the review, I have made the requested changes except for one where I left a comment. please take a look

mrakitin commented:
@awalter-bnl, I wanted to check the changes, but I saw there was a conflict in this PR, and now I don't see any of my suggestions committed in. I guess you are still resolving issues.

danielballan commented:
It was my fault @mrakitin. I squashed the commits since there were many syntax-error-related fixes, but I did a bad job of it. I believe the branch is now correct. Are there suggestions of yours left unaddressed?

danielballan commented:
Oh, I see your suggestions now. It should still be possible to apply them, I think.

mrakitin commented:
@danielballan, no, I don't see my changes. Normally, I see those suggested commits co-authored with the PR's author (@awalter-bnl in this case). No changes here, for example: https://github.com/bluesky/bluesky/pull/1190/files#r287093188.

awalter-bnl commented:
@mrakitin I did make all of your suggested changes, except for the one that you resolved because you agreed with my comment

danielballan commented:
> Normally, I see those suggested commits co-authored with the PR's author

Right. Any Suggested Changes committed by Andrew should have been enveloped in this squashed commit: b12d24b57d59eb1fee49edc4b47962781338fb1f

awalter-bnl commented:
So it turns out @mrakitin was correct, some of the commits got “undone” during the git surgery. I have now reapplies them.

danielballan commented:
Ha. I came to the same conclusion and went through and applied them manually (and some other instances of the same issues elsewhere in the code). Pushed in 528159258854ee39e84dc541ae0d7afa35567453.

awalter-bnl commented:
No problem, I am convinced they are in there now 😉

danielballan commented:
Taking a step back, I am very happy we are going to get this in. Just today, Eliot at SST emailed me with a use case that currently requires `scan_nd` and cyclers, but will be simpler with new-and-improved `list_scan`.

awalter-bnl commented:
Good to know, I suspect I can also get log scan done by tomorrow as well, even though it is far less useful 🙂

mrakitin commented on bluesky/plans.py:
```diff
@@ -143,19 +202,153 @@ def rel_list_scan(detectors, motor, steps, *, per_step=None, md=None):
+           'plan_name': 'list_grid_scan',
+           'plan_pattern': 'outer_list_product',
+           'plan_pattern_args': dict(args=md_args).update(
+                {'snake_axes':snake_axes}),
```
```suggestion
                {'snake_axes': snake_axes}),
```

mrakitin commented on bluesky/plans.py:
```diff
@@ -143,19 +202,153 @@ def rel_list_scan(detectors, motor, steps, *, per_step=None, md=None):
+def rel_list_grid_scan(detectors, *args, snake_axes=False, per_step=None,
+                       md=None):
+    """
+    Scan over a mesh; each motor is on an independent trajectory.Each point is
```
```suggestion
    Scan over a mesh; each motor is on an independent trajectory. Each point is
```

mrakitin commented on bluesky/plans.py:
```diff
@@ -143,19 +202,153 @@ def rel_list_scan(detectors, motor, steps, *, per_step=None, md=None):
+                       md=None):
+    """
+    Scan over a mesh; each motor is on an independent trajectory.Each point is
+    relative to the current postion.
```
```suggestion
    relative to the current position.
```

mrakitin commented on bluesky/plans.py:
```diff
@@ -143,19 +202,153 @@ def rel_list_scan(detectors, motor, steps, *, per_step=None, md=None):
+        simple left-to-right trajectory.
+
+    per_step : callable, optional
+        hook for customizing action of inner loop (messages per step)
```
```suggestion
        hook for customizing action of inner loop (messages per step).
```

mrakitin commented on bluesky/plans.py:
```diff
@@ -143,19 +202,153 @@ def rel_list_scan(detectors, motor, steps, *, per_step=None, md=None):
+        simple left-to-right trajectory.
+
+    per_step : callable, optional
+        hook for customizing action of inner loop (messages per step)
```
```suggestion
        hook for customizing action of inner loop (messages per step).
```

mrakitin commented on bluesky/plans.py:
```diff
@@ -676,8 +869,8 @@ def scan_nd(detectors, cycler, *, per_step=None, md=None):
         list of dictionaries mapping motors to positions
     per_step : callable, optional
         hook for customizing action of inner loop (messages per step)
```
```suggestion
        hook for customizing action of inner loop (messages per step).
```

mrakitin commented on bluesky/plans.py:
```diff
@@ -782,8 +975,8 @@ def scan(detectors, *args, num=None, per_step=None, md=None):
         number of points
     per_step : callable, optional
         hook for customizing action of inner loop (messages per step)
```
```suggestion
        hook for customizing action of inner loop (messages per step).
```

mrakitin commented on bluesky/plans.py:
```diff
@@ -869,8 +1062,8 @@ def grid_scan(detectors, *args, per_step=None, md=None):
         simple left-to-right trajectory.
     per_step : callable, optional
         hook for customizing action of inner loop (messages per step)
```
```suggestion
        hook for customizing action of inner loop (messages per step).
```

mrakitin commented on bluesky/plans.py:
```diff
@@ -941,8 +1134,8 @@ def rel_grid_scan(detectors, *args, per_step=None, md=None):
         trajectory or a simple left-to-right trajectory.
     per_step : callable, optional
         hook for customizing action of inner loop (messages per step)
```
```suggestion
        hook for customizing action of inner loop (messages per step).
```

mrakitin commented on bluesky/plans.py:
```diff
@@ -998,8 +1191,8 @@ def rel_scan(detectors, *args, num=None, per_step=None, md=None):
         number of points
     per_step : callable, optional
         hook for customizing action of inner loop (messages per step)
```
```suggestion
        hook for customizing action of inner loop (messages per step).
```

mrakitin commented on bluesky/plans.py:
```diff
@@ -1128,8 +1321,8 @@ def spiral_fermat(detectors, x_motor, y_motor, x_start, y_start, x_range,
         Tilt angle in radians, default 0.0
     per_step : callable, optional
         hook for customizing action of inner loop (messages per step)
```
```suggestion
        hook for customizing action of inner loop (messages per step).
```

mrakitin commented on bluesky/plans.py:
```diff
@@ -1197,8 +1390,8 @@ def rel_spiral_fermat(detectors, x_motor, y_motor, x_range, y_range, dr,
         Tilt angle in radians, default 0.0
     per_step : callable, optional
         hook for customizing action of inner loop (messages per step)
```
```suggestion
        hook for customizing action of inner loop (messages per step).
```

mrakitin commented on bluesky/plans.py:
```diff
@@ -1252,8 +1445,8 @@ def spiral(detectors, x_motor, y_motor, x_start, y_start, x_range, y_range, dr,
         Tilt angle in radians, default 0.0
     per_step : callable, optional
         hook for customizing action of inner loop (messages per step)
```
```suggestion
        hook for customizing action of inner loop (messages per step).
```

mrakitin commented on bluesky/plans.py:
```diff
@@ -1327,8 +1520,8 @@ def rel_spiral(detectors, x_motor, y_motor, x_range, y_range, dr, nth,
         Tilt angle in radians, default 0.0
     per_step : callable, optional
         hook for customizing action of inner loop (messages per step)
```
```suggestion
        hook for customizing action of inner loop (messages per step).
```

mrakitin commented on bluesky/plans.py:
```diff
@@ -1378,7 +1571,7 @@ def spiral_square(detectors, x_motor, y_motor, x_center, y_center, x_range,
         Number of y axis points.
     per_step : callable, optional
         hook for cutomizing action of inner loop (messages per step)
```
```suggestion
        hook for cutomizing action of inner loop (messages per step).
```

mrakitin commented on bluesky/plans.py:
```diff
@@ -1445,7 +1638,7 @@ def rel_spiral_square(detectors, x_motor, y_motor, x_range, y_range,
         Number of y axis points.
     per_step : callable, optional
         hook for cutomizing action of inner loop (messages per step)
```
```suggestion
        hook for cutomizing action of inner loop (messages per step).
```

mrakitin commented on bluesky/plans.py:
```diff
@@ -1609,8 +1802,8 @@ def x2x_scan(detectors, motor1, motor2, start, stop, num, *,
 
     per_step : callable, optional
         hook for cutomizing action of inner loop (messages per step)
```
```suggestion
        hook for cutomizing action of inner loop (messages per step).
```

awalter-bnl commented:
Thanks @mrakitin I have now made all of your suggestions, hopefully we found them all this time !-).
